File: implementations/typescript/packages/mcpproxy/src/schema.ts
// SPDX-License-Identifier: Apache-2.0
import { canonicalJson, record } from "@psp-cdl/core";

/** Deliberately finite JSON Schema subset. Unknown keywords are never ignored. */
export function checkSchema(schema:unknown, depth=0):void {
  if(!record(schema)||depth>16||typeof schema.type!=="string") throw new Error("UNSUPPORTED_SCHEMA");
  const fields:Record<string,string[]>={object:["properties","required","additionalProperties"],array:["items","minItems","maxItems"],string:["minLength","maxLength"],number:["minimum","maximum"],integer:["minimum","maximum"],boolean:[],null:[]};
  const type=schema.type;
  if(!Object.hasOwn(fields,type)||Object.keys(schema).some(k=>!["type","enum",...fields[type]!].includes(k))) throw new Error("UNSUPPORTED_SCHEMA");
  if("enum" in schema&&(!Array.isArray(schema.enum)||!schema.enum.length||schema.enum.length>1024)) throw new Error("UNSUPPORTED_SCHEMA");
  if(schema.type==="object") {
    if(!record(schema.properties)||schema.additionalProperties!==false||!Array.isArray(schema.required)||schema.required.some(k=>typeof k!=="string"||!Object.hasOwn(schema.properties as object,k))||new Set(schema.required).size!==schema.required.length) throw new Error("UNSUPPORTED_SCHEMA");
    for(const s of Object.values(schema.properties)) checkSchema(s,depth+1);
  }
  if(schema.type==="array") checkSchema(schema.items,depth+1);
  for(const [low,high] of [["minItems","maxItems"],["minLength","maxLength"],["minimum","maximum"]]) {
    for(const k of [low!,high!]) if(k in schema&&(typeof schema[k]!=="number"||!Number.isFinite(schema[k])||(k!=="minimum"&&k!=="maximum"&&(!Number.isSafeInteger(schema[k])||(schema[k] as number)<0)))) throw new Error("UNSUPPORTED_SCHEMA");
    if(low! in schema&&high! in schema&&(schema[low!] as number)>(schema[high!] as number)) throw new Error("UNSUPPORTED_SCHEMA");
  }
}

export function matches(schema:any,value:unknown):boolean {
  if(schema.enum&&!schema.enum.some((v:unknown)=>canonicalJson(v)===canonicalJson(value))) return false;
  const range=(n:number,lo:string,hi:string)=>(schema[lo]===undefined||n>=schema[lo])&&(schema[hi]===undefined||n<=schema[hi]);
  switch(schema.type) {
    case "object":return record(value)&&schema.required.every((k:string)=>Object.hasOwn(value,k))&&Object.keys(value).every(k=>Object.hasOwn(schema.properties,k)&&matches(schema.properties[k],value[k]));
    case "array":return Array.isArray(value)&&range(value.length,"minItems","maxItems")&&value.every(v=>matches(schema.items,v));
    case "string":return typeof value==="string"&&range([...value].length,"minLength","maxLength");
    case "number":case "integer":return typeof value==="number"&&Number.isFinite(value)&&(schema.type!=="integer"||Number.isInteger(value))&&range(value,"minimum","maximum");
    case "boolean":return typeof value==="boolean";
    case "null":return value===null;
    default:return false;
  }
}

File: implementations/python/packages/mcpproxy/src/psp_cdl_mcpproxy/schema.py
# SPDX-License-Identifier: Apache-2.0
"""Finite JSON Schema subset: unsupported keywords are never ignored."""
import math
from psp_cdl_core import canonical_json


def check_schema(schema, depth=0):
    fields = {"object":{"properties", "required", "additionalProperties"}, "array":{"items", "minItems", "maxItems"}, "string":{"minLength", "maxLength"}, "number":{"minimum", "maximum"}, "integer":{"minimum", "maximum"}, "boolean":set(), "null":set()}
    if type(schema) is not dict or depth > 16 or type(schema.get("type")) is not str or schema["type"] not in fields or set(schema) - ({"type", "enum"} | fields[schema["type"]]):
        raise ValueError("UNSUPPORTED_SCHEMA")
    if "enum" in schema and (type(schema["enum"]) is not list or not 1 <= len(schema["enum"]) <= 1024):
        raise ValueError("UNSUPPORTED_SCHEMA")
    if schema["type"] == "object":
        props, required = schema.get("properties"), schema.get("required")
        if type(props) is not dict or schema.get("additionalProperties") is not False or type(required) is not list or any(type(k) is not str or k not in props for k in required) or len(set(required)) != len(required):
            raise ValueError("UNSUPPORTED_SCHEMA")
        for child in props.values(): check_schema(child, depth + 1)
    if schema["type"] == "array": check_schema(schema.get("items"), depth + 1)
    for lo, hi in (("minItems", "maxItems"), ("minLength", "maxLength"), ("minimum", "maximum")):
        for k in (lo, hi):
            if k in schema:
                v = schema[k]
                if type(v) not in (int, float) or not math.isfinite(v) or (k not in ("minimum", "maximum") and (v < 0 or v > 9007199254740991 or int(v) != v)):
                    raise ValueError("UNSUPPORTED_SCHEMA")
        if lo in schema and hi in schema and schema[lo] > schema[hi]: raise ValueError("UNSUPPORTED_SCHEMA")


def matches(schema, value):
    if "enum" in schema and not any(canonical_json(v) == canonical_json(value) for v in schema["enum"]): return False
    def in_range(n, lo, hi): return (lo not in schema or n >= schema[lo]) and (hi not in schema or n <= schema[hi])
    kind = schema["type"]
    if kind == "object": return type(value) is dict and all(k in value for k in schema["required"]) and all(k in schema["properties"] and matches(schema["properties"][k], v) for k, v in value.items())
    if kind == "array": return type(value) is list and in_range(len(value), "minItems", "maxItems") and all(matches(schema["items"], v) for v in value)
    if kind == "string": return type(value) is str and in_range(len(value), "minLength", "maxLength")
    if kind in ("number", "integer"): return type(value) in (int, float) and math.isfinite(value) and (kind != "integer" or int(value) == value) and in_range(value, "minimum", "maximum")
    if kind == "boolean": return type(value) is bool
    if kind == "null": return value is None
    return False
